File: api/webhooks.py
import asyncio

from fastapi import APIRouter, BackgroundTasks, HTTPException, Request

from agents.workflow import AgentState, workflow
from store import RepoMappingStore

router = APIRouter(prefix="/webhooks", tags=["webhooks"])


@router.post("/jira")
async def jira_webhook(request: Request, background_tasks: BackgroundTasks):
    payload = await request.json()
    issue = payload.get("issue")
    if not issue:
        event_type = payload.get("issue_event_type_name")
        if event_type:
            issue = payload
        else:
            raise HTTPException(status_code=400, detail="No issue data in payload")

    issue_key = issue.get("key")
    fields = issue.get("fields", {})
    project_key = fields.get("project", {}).get("key")

    if not issue_key or not project_key:
        raise HTTPException(status_code=400, detail="Missing issue key or project key")

    mapping = RepoMappingStore.get_by_project(project_key)
    if not mapping:
        raise HTTPException(
            status_code=404, detail=f"No repo mapping for project {project_key}"
        )

    repo_url = f"https://github.com/{mapping['github_repo']}.git"

    initial_state = AgentState(
        ticket_id=issue_key,
        ticket_details={},
        repo_url=repo_url,
        local_path="",
        branch_name="",
        repo_full_name=mapping["github_repo"],
        base_branch=mapping.get("base_branch", "main"),
        code_diff=[],
        pr_url="",
        error="",
    )

    background_tasks.add_task(_run_workflow_sync, initial_state)
    return {"status": "accepted", "ticket_id": issue_key}


def _run_workflow_sync(state: AgentState):
    asyncio.run(workflow.ainvoke(state))

File: src/api/webhooks.ts
import type { FastifyInstance } from "fastify";
import { RepoMappingStore } from "../store/mappings.js";
import { AgentEngine } from "../agent/engine.js";
import { logger } from "../utils/logger.js";
import type { JiraWebhookPayload } from "../types/index.js";

export async function webhookRoutes(fastify: FastifyInstance): Promise<void> {
  const agentEngine = new AgentEngine();
  await agentEngine.initialize();

  fastify.addHook("onClose", async () => {
    await agentEngine.shutdown();
  });

  fastify.post<{
    Body: JiraWebhookPayload;
  }>("/jira", async (request, reply) => {
    const payload = request.body;

    // Extract issue data
    let issue = payload.issue;
    if (!issue && payload.issue_event_type_name) {
      // Some JIRA webhooks wrap the issue differently
      issue = payload as unknown as JiraWebhookPayload["issue"];
    }

    if (!issue) {
      return reply.status(400).send({ error: "No issue data in payload" });
    }

    const issueKey = issue.key;
    const projectKey = issue.fields?.project?.key;

    if (!issueKey || !projectKey) {
      return reply.status(400).send({ error: "Missing issue key or project key" });
    }

    // Lookup repo mapping
    const mapping = await RepoMappingStore.getByProject(projectKey);
    if (!mapping) {
      return reply.status(404).send({ error: `No repo mapping for project ${projectKey}` });
    }

    // Build ticket from webhook payload to avoid redundant JIRA fetch
    const ticket = {
      id: issue.id || "",
      key: issueKey,
      summary: issue.fields.summary || "",
      description: issue.fields.description || null,
      status: issue.fields.status?.name || "Unknown",
      issueType: issue.fields.issuetype?.name || "Unknown",
    };

    // Process in background (don't await)
    agentEngine
      .processTicket(issueKey, projectKey, ticket)
      .then((result) => {
        if (result.success) {
          logger.info({ issueKey }, "Ticket processed successfully");
        } else {
          logger.error({ issueKey, error: result.error }, "Ticket processing failed");
        }
      })
      .catch((error) => {
        logger.error({ issueKey, error }, "Unexpected error processing ticket");
      });

    return reply.status(202).send({ status: "accepted", ticket_id: issueKey });
  });
}
